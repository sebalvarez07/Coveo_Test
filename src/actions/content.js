import axios from 'axios';
import { token } from '../fixtures/helpers';
import { groupMaker, queryGenerator, parseBaseFilter } from '../fixtures/query';

export const setContent = (content) => {
    return {
        type: 'SET_CONTENT',
        content
    }
}

export const startSetContent = () => {
    return (dispatch, getState) => {
        
        const filters = getState().filters;

        const fullQuery = queryGenerator(filters);

        return axios.post(`https://cloudplatform.coveo.com/rest/search?access_token=${token}`, {
            // Send query
            ...fullQuery
            })
            .then(function (response) {
                // console.log('response', response);
                dispatch(setContent({
                    groupBy: response.data.groupByResults,
                    results: response.data.results,
                    resultsCount: response.data.totalCountFiltered
                }));
            })
            .catch(function (error) {
                console.log(error);
            }
        );
    }
}

export const updateCategoryLimit_Content = (group, groupValues) => {
    return {
        type: 'UPDATE_GROUP_LIMIT',
        group,
        groupValues
    }
}

export const startUpdateCategoryLimit_Content = (group, valueLimit) => {
    return (dispatch, getState) => {

        const baseFilter = parseBaseFilter(getState().filters.baseFilter);
        const groupUpdated = groupMaker(group, valueLimit, baseFilter);
        const query = {
            "groupBy": [
                { ...groupUpdated }
            ]
        }
        // console.log('query limit', query);
        return axios.post(`https://cloudplatform.coveo.com/rest/search?access_token=${token}`, {
            ...query
        }).then(result => {
            const groupValues = result.data.groupByResults[0];
            dispatch(updateCategoryLimit_Content(group, groupValues))
        })
    }
}