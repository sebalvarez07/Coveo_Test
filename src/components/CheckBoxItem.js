import React, { useState } from 'react';

const CheckBoxItem = (props) => {

};

export default CheckBoxItem;