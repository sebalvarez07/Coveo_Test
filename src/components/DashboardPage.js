import React from 'react';
import Header from './Header';
import Sidebar from './Sidebar';
import MainContainer from './MainContainer';
 
const DashboardPage = (props) => {
    return (
        <div className='wrapper'>
            <Header />
            <Sidebar />
            <MainContainer />
        </div>
    )
}

export default DashboardPage;