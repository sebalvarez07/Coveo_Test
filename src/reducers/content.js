const contentReducer = (state = {}, action) => {
    switch(action.type) {
        case 'SET_CONTENT':
            return action.content;

        case 'UPDATE_GROUP_LIMIT':
            const updatedGroup = state.groupBy.map(group => {
                if(group.field === action.group) {
                    return action.groupValues
                }
                return group
            })
            return {
                ...state,
                groupBy: updatedGroup,
            }
        default:
            return state;
    }
}

export default contentReducer;