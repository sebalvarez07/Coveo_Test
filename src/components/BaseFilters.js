import React from 'react';
import { connect } from 'react-redux';
import { startUpdateBaseFilter } from '../actions/filters';

export class BaseFilters extends React.Component {

    constructor (props) {
        super(props);
        this.beerButton = React.createRef();
        this.wineButton = React.createRef();
        this.state = {
            beerActive: true,
            wineActive: true
        }
    }

    componentDidUpdate (prevProps, prevState) {
        if(prevState.beerActive !== this.state.beerActive) {
            this.beerButton.current.classList.toggle('active-base');
        }
        
        if(prevState.wineActive !== this.state.wineActive) {
            this.wineButton.current.classList.toggle('active-base');
        }
    }
    
    dispatchBaseFilter = () => {
        let baseFilter;
        const beerActive = this.state.beerActive;
        const wineActive = this.state.wineActive;

        if(beerActive === wineActive) {
            baseFilter = ['Bière', 'Vin'];
        }
        else if(beerActive && !wineActive) {
            baseFilter = ['Bière'];
        }
        else if(!beerActive && wineActive) {
            baseFilter = ['Vin'];
        }

        this.props.startUpdateBaseFilter(baseFilter);
    }

    handleBaseFilter = (e) => {
        e.preventDefault(); 
        const btn = e.target.closest('.base-filter__btn');

        if(btn.id === 'beerFilter') {
            this.setState(prevState => (
                {
                    beerActive: !prevState.beerActive
                }
                ), this.dispatchBaseFilter);
        }
        else {
            this.setState(prevState => (
                {
                    wineActive: !prevState.wineActive
                }
                ), this.dispatchBaseFilter);
        }
    }

    render () {
        return (
            <div className='base-filter header__el'>
                <span 
                    ref={this.beerButton}
                    className='base-filter__btn active-base'
                    id={'beerFilter'}
                    onMouseDown={this.handleBaseFilter}
                >
                    <i className="ionicons ion-ios-pint"></i>
                    <span className='base-filter__label'>
                        Beer
                    </span>
                </span>
                <span
                    ref={this.wineButton}
                    className='base-filter__btn active-base'
                    id={'wineFilter'}
                    onMouseDown={this.handleBaseFilter}
                >
                    <i className="ionicons ion-ios-wineglass"></i>
                    <span className='base-filter__label'>
                        Wine
                    </span>
                </span>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startUpdateBaseFilter: (baseFilter) => dispatch(startUpdateBaseFilter(baseFilter))
    }
}
 
export default connect(undefined, mapDispatchToProps)(BaseFilters);