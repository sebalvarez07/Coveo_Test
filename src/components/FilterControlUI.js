import React from 'react';
import { connect } from 'react-redux';
import { groupMap } from '../fixtures/helpers';
import { startRemoveFilterValue } from '../actions/filters';

export const FilterControlUI = (props) => {

    const handleValueRemoval = (e) => {
        const btn = e.target.closest('.filter-control__btn');
        const group = btn.dataset.group;
        const val = btn.dataset.value;
        props.startRemoveFilterValue(group, val);
    } 

    return (
        <div className='filter-controlUI'>
            <h4 className='current-viewing'>
                Currently Viewing: {props.baseFilter.join(' & ')}
            </h4>
            {
                Object.keys(props.filtersActive).map(category => {
                    if(props.filtersActive[`${category}`].values.length > 0) {
                        return (
                            <div className='filter-control__container' key={category}>
                                <span className='filter-control__cat'>{groupMap.get(category)}: </span>
                                {
                                    props.filtersActive[category].values.map(activeFilterVal => {
                                        return <button 
                                                    data-group={category}
                                                    data-value={activeFilterVal}
                                                    className='filter-control__btn'
                                                    onMouseDown={handleValueRemoval}
                                                    key={activeFilterVal}>
                                                        {activeFilterVal}
                                                        <i className="ionicons ion-close"></i>
                                                </button>
                                    })
                                }
                            </div>
                        )    
                    }
                })
            }
        </div>
    )
}

const mapStateToProps = (state) => (
    {
        filtersActive: state.filters.categories,
        baseFilter: state.filters.baseFilter
    }
);

const mapDispatchToProps = (dispatch) => (
    {
        startRemoveFilterValue: (group, value) => dispatch(startRemoveFilterValue(group, value)),
    }
);

export default connect(mapStateToProps, mapDispatchToProps)(FilterControlUI);