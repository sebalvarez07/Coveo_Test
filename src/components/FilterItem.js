import React from 'react';
import { connect } from 'react-redux';
import { startUpdateCategory, startUpdateCategoryLimit } from'../actions/filters';
import { groupMap } from '../fixtures/helpers';

const defaultLimit = 5;

class FilterItem extends React.Component {

    constructor(props) {
        super(props);
        this.checkBoxes = [];
        this.checkedItems = new Set();
        this.state = {
            valueLimit: defaultLimit
        }
    }

    componentWillMount () {
        this.checkBoxes = this.props.filter.map(option => option.value);
    }

    componentWillUpdate (nextProps) {

        this.checkedItems = new Set([...this.props.checkedFilters]);

        if(nextProps.baseFilter !== this.props.baseFilter) {
            // TODO.. on baseFilter change, remove wine || beer filters           
        }
        
        // Ensure to keep check boxes on view
        this.checkBoxes = [...this.checkedItems];
    
        // Keep old checked boxes, and add new boxes
        nextProps.filter.forEach(option => {
            if(!this.checkedItems.has(option.value)){
                this.checkBoxes.push(option.value)
            }
        });     
    }

    handleChange = (e) => {
        const item = e.target.name;
        const isChecked = e.target.checked;
            
        if(isChecked) {
            this.checkedItems.add(item)
        } else {
            if(this.checkedItems.has(item)) {
                this.checkedItems.delete(item);
            }
        }

        const checkedItemsArr = Array.from(this.checkedItems);

        this.props.startUpdateCategory(this.props.filterName, checkedItemsArr);
    }

    dispatchMoreValues = () => {
        if(this.state.valueLimit <= this.props.filter.length) {
            return;
        } else {
            this.props.startUpdateCategoryLimit(this.props.filterName, this.state.valueLimit);
        }
    }

    handleValueLimit = (e) => {
        e.preventDefault();
        
        if(e.target.name === 'more'){
            this.setState(prevState => {
                return {
                    valueLimit: prevState.valueLimit + 5
                }
            }, this.dispatchMoreValues)
        }

        else if(e.target.name === 'less'){
            const resetLimit = this.checkedItems.size > defaultLimit ? this.checkedItems.size : defaultLimit;
            this.setState({ valueLimit: resetLimit});
        }
    }

    handleClearAll = () => {
        this.props.startUpdateCategory(this.props.filterName, []);
    }

    setLimit = () => {
        
        return this.checkedItems.size > this.state.valueLimit ? this.checkedItems.size : this.state.valueLimit;
    }

    renderCheckBox = (checkName) => {        
        return (
            <label className='check__container' key={`${checkName}`}>
                {`${checkName}`}
                <input 
                    className='custom__checkbox'
                    style={{ display: 'block' }} 
                    type='checkbox' 
                    name={`${checkName}`} 
                    checked={this.checkedItems.has(checkName)}
                    onChange={this.handleChange} 
                />
                <span className='checkmark'>
                    <i className="ionicons ion-checkmark"></i>
                </span>
            </label>
        )
    }
    
    render (){
        const limit = this.setLimit();
        return (
            <div className='sidebar__group'>
                <div className='group__header'>
                    <h3 className='group__tl'>{groupMap.get(this.props.filterName)}</h3>
                    <span className='clear-btn' onMouseDown={this.handleClearAll}>Clear All</span>
                </div>
                
                {
                    this.checkBoxes.reduce((acc, checkName, index) => {
                        if(limit > index) {
                            acc.push(this.renderCheckBox(checkName))
                        }
                        return acc;
                    }, [])
                }
                
                <div className='more_less__group'>
                    { this.state.valueLimit <= this.props.filter.length && 
                        <React.Fragment>
                            <button 
                                className='more_less__btn'
                                name={'more'}
                                onMouseDown={this.handleValueLimit}
                                >
                                Show More
                            </button>
                            <span> | </span>
                        </React.Fragment>
                    }
                    { this.props.filter.length >= defaultLimit && 
                        <button 
                            className='more_less__btn'
                            name={'less'}
                            onMouseDown={this.handleValueLimit}
                            >
                            Show Less
                        </button>
                    }
                </div>
                
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startUpdateCategory: (group, update) => dispatch(startUpdateCategory(group, update)),
        startUpdateCategoryLimit: (group, valueLimit) => dispatch(startUpdateCategoryLimit(group, valueLimit))
    }
}

const mapStateToProps = (state, props) => {
    return  {
        baseFilter: state.filters.baseFilter,
        checkedFilters: state.filters.categories[`${props.filterName}`]['values']
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(FilterItem);