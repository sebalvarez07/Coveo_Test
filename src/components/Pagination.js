import React from 'react';
import { connect } from 'react-redux';
import { startUpdatePageNum } from '../actions/filters';
 
export const Pagination = (props) => {

    // Set all pagination variables 
    const numberOfPages = Math.ceil(props.totalResults / props.numberOfResults);
    const currentPage = Math.ceil(props.firstResult / props.numberOfResults) + 1;
    const limitPages = numberOfPages < 5 ? numberOfPages : 5;
    
    const setPages = () => {
        let startingPage;
        
        if(currentPage < 4 || numberOfPages < 5) {
            startingPage = 1
        } else if(currentPage >= numberOfPages - 2) {
            startingPage = (currentPage - 4) + (numberOfPages - currentPage)
        } else  {
            // Start with offset if current page is at 4 or higher
            startingPage = currentPage - 2
        }

        let temp = [];
        for(let i = startingPage; i < startingPage + limitPages; i++) {
            temp.push(i);
        }

        return temp;
    }

    const handlePageNumber = (e) => {
        const dataTarget = e.target.dataset.target;
        const calcNextResult = (dataTarget - 1) * props.numberOfResults;
        // Since API doesn't return a number higher than 1000, we go back to 1
        const nextFirstResult = calcNextResult >= 1000 ? 0 : calcNextResult;
        props.startUpdatePageNum(nextFirstResult);
    }
    
    // Button component for all pagination buttons
    const PaginationButton = ({pageNum, displayVal = pageNum, className, onMouseDown}) => {
        return (
            <button
                className={`pagination__btn ${className} ${pageNum === currentPage ? 'active-page' : ''}`}
                onMouseDown={onMouseDown}  
                data-target={pageNum}
            >
                {displayVal}
            </button>
        )
    }

    // Populate array with page numbers
    let pages = setPages();

    return (
        <div className='pagination'>
            {
                currentPage > 1 &&
                <PaginationButton onMouseDown={handlePageNumber} pageNum={currentPage - 1} displayVal={'Prev'} className={'next-prev-btn prev-btn'}/>
            }
            {
                (currentPage > 4) &&
                <PaginationButton onMouseDown={handlePageNumber} pageNum={1} displayVal={'1...'} className={'next-prev-btn pagination-1'}/>
            }
            {
                numberOfPages > 1 &&
                pages.map(page => {
                  return <PaginationButton onMouseDown={handlePageNumber} key={page} pageNum={page} className={`page-btn pagination-${page}`}/>
                })
            }
            {
                currentPage < numberOfPages &&
                <PaginationButton onMouseDown={handlePageNumber} pageNum={currentPage + 1} displayVal={'Next'} className={'next-prev-btn next-btn'}/>
            }
        </div>
    )
}
 
const mapStateToProps = (state) => (
    {
        totalResults: state.content.resultsCount,
        numberOfResults: state.filters.numberOfResults,
        firstResult: state.filters.firstResult
    }
);

const mapDispatchToProps = (dispatch) => (
    {
        startUpdatePageNum: (firstResult) => dispatch(startUpdatePageNum(firstResult))
    }
);

export default connect(mapStateToProps, mapDispatchToProps)(Pagination);