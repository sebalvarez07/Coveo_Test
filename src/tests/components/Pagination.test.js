import React from 'react';
import { shallow } from 'enzyme';
import { Pagination } from '../../components/Pagination';
import { contentEx } from '../fixtures/content';

let wrapper, totalResults, numberOfResults, firstResult, startUpdatePageNum;

beforeEach( () => {
    totalResults = 100;
    numberOfResults = 20;
    firstResult = 40;
    startUpdatePageNum = jest.fn();
    wrapper = shallow(<Pagination 
                            totalResults={totalResults} 
                            numberOfResults={numberOfResults}
                            firstResult={firstResult}
                            startUpdatePageNum={startUpdatePageNum}
                            />)
});

test('Should render base filters', () => {
    expect(wrapper).toMatchSnapshot();
}); 

test('Should handle startUpdatePageNum when requesting the next page', () => {
    
    const currentPage = Math.ceil(firstResult / numberOfResults) + 1;
    // the equations is target = currentPage + 1 BUT later we do dataTarget - 1 (to get the right index)
    const target = currentPage + 1;     
    const nextFirstResult = (target - 1) * numberOfResults;

    wrapper.find('.next-btn').prop('onMouseDown')({
        target: {
           dataset: { target }
        }
    });

    expect(startUpdatePageNum).toHaveBeenLastCalledWith(nextFirstResult);
});


test('Should handle startUpdatePageNum when requesting the previous page', () => {
    
    const currentPage = Math.ceil(firstResult / numberOfResults) + 1;
    // the equations is target = currentPage + 1 BUT later we do dataTarget - 1 (to get the right index)
    const target = currentPage - 1;     
    const nextFirstResult = (target - 1) * numberOfResults;

    wrapper.find('.prev-btn').prop('onMouseDown')({
        target: {
           dataset: { target }
        }
    });

    expect(startUpdatePageNum).toHaveBeenLastCalledWith(nextFirstResult);
});

test('Should handle startUpdatePageNum when requesting any page', () => {
    
    // the equations is target = currentPage + 1 BUT later we do dataTarget - 1 (to get the right index)
    const target = 3;     
    const nextFirstResult = (target - 1) * numberOfResults;

    wrapper.find('.pagination-3').prop('onMouseDown')({
        target: {
           dataset: { target }
        }
    });

    expect(startUpdatePageNum).toHaveBeenLastCalledWith(nextFirstResult);
});