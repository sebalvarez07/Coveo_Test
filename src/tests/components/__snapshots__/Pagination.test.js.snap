// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render base filters 1`] = `
<div
  className="pagination"
>
  <PaginationButton
    className="next-prev-btn prev-btn"
    displayVal="Prev"
    onMouseDown={[Function]}
    pageNum={2}
  />
  <PaginationButton
    className="page-btn pagination-1"
    key="1"
    onMouseDown={[Function]}
    pageNum={1}
  />
  <PaginationButton
    className="page-btn pagination-2"
    key="2"
    onMouseDown={[Function]}
    pageNum={2}
  />
  <PaginationButton
    className="page-btn pagination-3"
    key="3"
    onMouseDown={[Function]}
    pageNum={3}
  />
  <PaginationButton
    className="page-btn pagination-4"
    key="4"
    onMouseDown={[Function]}
    pageNum={4}
  />
  <PaginationButton
    className="page-btn pagination-5"
    key="5"
    onMouseDown={[Function]}
    pageNum={5}
  />
  <PaginationButton
    className="next-prev-btn next-btn"
    displayVal="Next"
    onMouseDown={[Function]}
    pageNum={4}
  />
</div>
`;
